shared/controlflow/codeql/controlflow/Cfg.qll

```diff
@@ -1514,6 +1528,260 @@ module MakeWithSplitting<
     /** Holds if CFG scope `scope` lacks an initial AST node. */
     query predicate scopeNoFirst(CfgScope scope) { not scopeFirst(scope, _) }
   }
+
+  /**
+   * Provides a basic block construction on top of the control flow graph.
+   */
+  module BasicBlocks {
+    private import codeql.controlflow.BasicBlock as BB
+
+    private class NodeAlias = Node;
+
+    private module BasicBlockInputSig implements BB::InputSig<Location> {
+      class SuccessorType = Input::SuccessorType;
+
+      predicate successorTypeIsCondition = Input::successorTypeIsCondition/1;
+
+      class Node = NodeAlias;
+
+      Node nodeGetASuccessor(Node node, SuccessorType t) { result = node.getASuccessor(t) }
+
+      predicate nodeIsDominanceEntry(Node node) { node instanceof EntryNode }
+
+      predicate nodeIsPostDominanceExit(Node node) { node.(AnnotatedExitNode).isNormal() }
+    }
+
+    private module BasicBlockImpl = BB::Make<Location, BasicBlockInputSig>;
+
+    /**
+     * A basic block, that is, a maximal straight-line sequence of control flow nodes
+     * without branches or joins.
+     */
+    final class BasicBlock extends BasicBlockImpl::BasicBlock {
+      // We extend `BasicBlockImpl::BasicBlock` to add the `getScope`.
+      /** Gets the scope of this basic block. */
+      CfgScope getScope() {
```

Comment:
Thoughts on putting `Node.getScope` through the BasicBlock input, such that `getScope` could be added to BasicBlock directly in its definition inside `shared/controlflow/codeql/controlflow/BasicBlock.qll`? That way this could just be an alias rather than having to copy all the member predicates.

Reply:
It would definitely save some boiler plate, but could make the `BasicBlock` module less flexible for instantiations that don't want a `getScope`.

Reply:
What do you think about this @hvitved?

Reply:
Makes sense to me.